### PR TITLE
RSDK-3016 cache the frame system config

### PR DESCRIPTION
**Pull request by raybjork**
This work makes it so that we don't have to rebuild the frame system config every time it is requested which will result in fewer queries to remotes

### CODE REVIEW

File: robot/impl/local_robot.go [hunk 1]
```diff
@@ -724,10 +730,17 @@ func (r *localRobot) updateWeakDependents(ctx context.Context) {
 	}
 }
 
-// Config returns the info of each individual part that makes up the frame system
+// config returns the info of each individual part that makes up the frame system
 // The output of this function is to be sent over GRPC to the client, so the client
 // can build its frame system. requests the remote components from the remote's frame system service.
 func (r *localRobot) FrameSystemConfig(ctx context.Context) (*framesystem.Config, error) {
+	if r.fsConfig == nil {
```
Comment (edaniels) [inline robot/impl/local_robot.go:737]: I think this needs a lock right?

File: robot/impl/local_robot.go [hunk 2]
```diff
@@ -45,6 +45,9 @@ type localRobot struct {
 	manager *resourceManager
 	config  *config.Config
 
+	fsConfig    *framesystem.Config
+	lastFSError error
```
Comment (raybjork) [inline robot/impl/local_robot.go:49]: it jumps out to me as a little weird to be storing this in the robot struct.  open to other ideas

Comment (edaniels) [reply robot/impl/local_robot.go:49]: it seems like framesystem service can hold this

**Review by biotinker — COMMENTED** [review]
Could you please test this and verify that it behaves properly on a local when a remote robot also running this branch has its frame system changed?

As we all know, there are only two hard problems in computer science: cache invalidation, naming things, and off-by-1 errors.

File: robot/impl/local_robot.go (hunk 1 above)
Comment (raybjork) [reply robot/impl/local_robot.go:737]: yes good call

File: robot/impl/local_robot.go (hunk 2 above)
Comment (raybjork) [reply robot/impl/local_robot.go:49]: Yeah I toyed with this, but some of the errors we care about currently exist in the robot file (i.e. `errors.Errorf("cannot give frame system part the name %s", referenceframe.World)`

A way around this would be removing all checks from the robot level and passing in whatever config we get for the frame system service to handle.  This is probably the better approach I'll look into this

Comment (biotinker) [reply robot/impl/local_robot.go:49]: Agree this is the right approach here

File: components/arm/client_test.go [hunk 3]
```diff
@@ -110,8 +109,8 @@ func TestClient(t *testing.T) {
 	test.That(t, resourceAPI.RegisterRPCService(context.Background(), rpcServer, armSvc), test.ShouldBeNil)
 
 	injectRobot := &inject.Robot{}
-	injectRobot.FrameSystemConfigFunc = func(ctx context.Context) (*framesystem.Config, error) {
-		return &framesystem.Config{}, nil
+	injectRobot.FrameSystemConfigFunc = func(ctx context.Context) ([]*referenceframe.FrameSystemPart, error) {
```
Comment (raybjork) [inline components/arm/client_test.go:112]: This no longer returns a config and therefore maybe should be renamed something like `FrameSystem`

File: components/camera/videosource/join_pointcloud_test.go [hunk 4]
```diff
@@ -69,22 +69,31 @@ func makeFakeRobot(t *testing.T) resource.Dependencies {
 	}
 	base1 := &inject.Base{}
 
-	fsParts := []*referenceframe.FrameSystemPart{
+	fsParts := []*framesystem.PartConfig{
 		{
-			FrameConfig: referenceframe.NewLinkInFrame(referenceframe.World, spatialmath.NewZeroPose(), "base1", nil),
```
Comment (raybjork) [inline components/camera/videosource/join_pointcloud_test.go:74]: This name is also a little outdated, and should probably read `Frame` instead

File: referenceframe/frame_json.go [hunk 5]
```diff
@@ -87,6 +94,18 @@ func (cfg *LinkConfig) ParseConfig() (*LinkInFrame, error) {
 	return NewLinkInFrame(cfg.Parent, pose, cfg.ID, geom), nil
 }
 
+// Rename makes a deep copy of the config with an ID corresponding to the given name.
+func (cfg *LinkConfig) Rename(name string) *LinkConfig {
```
Comment (biotinker) [inline referenceframe/frame_json.go:98]: `Rename` is a misleading name here; it makes one think they're mutating the underlying struct.

Would prefer something like `CopyWithNewName`

File: robot/impl/local_robot.go [hunk 6]
```diff
@@ -735,112 +735,67 @@ func (r *localRobot) updateWeakDependents(ctx context.Context) {
...
+// config returns the info of each individual part that makes up the frame system
 // The output of this function is to be sent over GRPC to the client, so the client
 // can build its frame system. requests the remote components from the remote's frame system service.
-func (r *localRobot) FrameSystemConfig(ctx context.Context) (*framesystem.Config, error) {
-	localParts, err := r.getLocalFrameSystemParts()
-	if err != nil {
-		return nil, err
-	}
-	remoteParts, err := r.getRemoteFrameSystemParts(ctx)
+func (r *localRobot) FrameSystemConfig(ctx context.Context) ([]*referenceframe.FrameSystemPart, error) {
```
Comment (benjirewis) [inline robot/impl/local_robot.go:741]: Can we rename this and the associated inject function something like `FrameSystemParts`?

File: robot/impl/local_robot.go [hunk 7]
```diff
@@ -735,112 +735,67 @@ func (r *localRobot) updateWeakDependents(ctx context.Context) {
 	}
 }
 
-// Config returns the info of each individual part that makes up the frame system
+// config returns the info of each individual part that makes up the frame system
```
Comment (benjirewis) [inline robot/impl/local_robot.go:738]:
```suggestion
// FrameSystemParts returns the info of each individual part that makes up the frame system
```

File: robot/impl/local_robot.go [hunk 8]
```diff
@@ -690,7 +690,7 @@ func (r *localRobot) updateWeakDependents(ctx context.Context) {
 				r.Logger().Errorw("failed to reconfigure internal service", "service", resName, "error", err)
 			}
 		case framesystem.InternalServiceName:
-			fsCfg, err := r.FrameSystemConfig(ctx)
+			fsCfg, err := r.buildFrameSystemConfig(ctx)
```
Comment (benjirewis) [inline robot/impl/local_robot.go:693]: To clarify, you don't want to use cached parts here?

File: robot/client/client_test.go [hunk 9]
```diff
@@ -121,17 +120,17 @@ func TestStatusClient(t *testing.T) {
 	}
 
 	// TODO(RSDK-882): will update this so that this is not necessary
```
Comment (benjirewis) [inline robot/client/client_test.go:122]: RSDK-882 is closed, what's the status of this `TODO`?

File: robot/framesystem/framesystem.go [hunk 10]
```diff
@@ -74,38 +75,20 @@ func (svc *frameSystemService) Name() resource.Name {
...
-				utils.RadToDeg(ori.Roll),
-				utils.RadToDeg(ori.Pitch),
-				utils.RadToDeg(ori.Yaw),
-			),
-			geomString,
-		})
-	}
-	return t.Render()
+// PartConfig is the structure that encodes all the information needed for a frame system part.
+// Only one of the two fields between FrameConfig and PreprossedPart should be non-nil as they each provide the same
```
Comment (benjirewis) [inline robot/framesystem/framesystem.go:82]:
```suggestion
// Only one of the two fields between FrameConfig and PreprocessedPart should be non-nil as they each provide the same
```

File: robot/framesystem/framesystem.go [hunk 11]
```diff
@@ -131,25 +143,38 @@ func (svc *frameSystemService) Reconfigure(ctx context.Context, deps resource.De
...
+		svc.parts = nil
+		svc.lastReconfigureError = err
 		return err
 	}
 
-	sortedParts, err := referenceframe.TopologicallySortParts(fsCfg.Parts)
+	parts, err := svc.ConfigToParts(fsCfg)
 	if err != nil {
+		svc.parts = nil
+		svc.lastReconfigureError = err
```
Comment (benjirewis) [inline robot/framesystem/framesystem.go:165]: Why do you need to cache the reconfiguration error? Is returning it from `Reconfigure` not enough?

File: robot/impl/data/fake_wrongconfig4.json [hunk 12]
```diff
@@ -11,7 +11,7 @@
         {
             "name": "cameraOver",
             "type": "camera",
-            "model": "file",
+            "model": "fake",
```
Comment (benjirewis) [inline robot/impl/data/fake_wrongconfig4.json:14]: ?

**Review by benjirewis — CHANGES_REQUESTED** [review]

File: referenceframe/frame_json.go [hunk 13]
```diff
@@ -70,6 +71,12 @@ func NewLinkConfig(frame staticFrame) (*LinkConfig, error) {
 
 // ParseConfig converts a LinkConfig into a staticFrame.
 func (cfg *LinkConfig) ParseConfig() (*LinkInFrame, error) {
+	if cfg.ID == World {
+		return nil, errors.Errorf("cannot give frame system part the name %s", World)
+	}
+	if cfg.Parent == "" {
+		return nil, errors.Errorf("parent field in frame config for part %q is empty", cfg.ID)
```
Comment (benjirewis) [inline referenceframe/frame_json.go:78]: I think this new check is causing `TestSerializationStatic` to fail.

File: robot/impl/local_robot.go (hunk 8 above)
Comment (raybjork) [reply robot/impl/local_robot.go:693]: I might be misunderstanding something here, but I think during an update to weak dependence we would want to update what is being cached in the frame system.  This caching is useful for when we are looking to consume the framesystem

File: robot/client/client_test.go (hunk 9 above)
Comment (raybjork) [reply robot/client/client_test.go:122]: I'm not sure.  Theres 7 instances of this TODO in tests.  I'm going to try to remove the offending code and see if it breaks now.

File: robot/framesystem/framesystem.go (hunk 11 above)
Comment (raybjork) [reply robot/framesystem/framesystem.go:165]: The reason for this update is that we often call the function soon to be named `FrameSystemParts` from the motion service (as well as in a couple other places). If we try to do that on a framesystem service that has failed reconfigure we would have to error out but wouldn't know why since we wouldn't be rebuilding it anymore.  Rather than returning something generic I would rather return the reason why it failed reconfigure

File: robot/impl/data/fake_wrongconfig4.json (hunk 12 above)
Comment (raybjork) [reply robot/impl/data/fake_wrongconfig4.json:14]: I changed this because it was causing issues in testing.  Khari said it got renamed to "image_file" so I'll rename to that

File: robot/impl/local_robot.go (hunk 8 above)
Comment (benjirewis) [reply robot/impl/local_robot.go:693]: Cool, that sounds correct to me.

File: robot/client/client_test.go (hunk 9 above)
Comment (raybjork) [reply robot/client/client_test.go]: Nice catch! This was made totally unnecessary by recent frame system changes.  Removed.

**Review by benjirewis — APPROVED** [review]
LGTM. Great to remove all the RSDK-882 stuff, thanks 🧑‍🔧 .

File: referenceframe/frame_json.go [hunk 14]
```diff
@@ -70,6 +71,12 @@ func NewLinkConfig(frame staticFrame) (*LinkConfig, error) {
 
 // ParseConfig converts a LinkConfig into a staticFrame.
 func (cfg *LinkConfig) ParseConfig() (*LinkInFrame, error) {
+	if cfg.ID == World {
```
Comment (biotinker) [inline referenceframe/frame_json.go:74]: Will this break when receiving the framesystem from a remote robot's FS? Can we test this?

File: robot/framesystem/framesystem.go [hunk 15]
```diff
@@ -74,38 +75,20 @@ func (svc *frameSystemService) Name() resource.Name {
 // Config is a slice of *config.FrameSystemPart.
 type Config struct {
 	resource.TriviallyValidateConfig
-	Parts                []*referenceframe.FrameSystemPart
-	AdditionalTransforms []*referenceframe.LinkInFrame
+	PartConfigs []*PartConfig
 }
 
-// String prints out a table of each frame in the system, with columns of name, parent, translation and orientation.
```
Comment (biotinker) [inline robot/framesystem/framesystem.go:81]: The prettyprint is useful; why are we getting rid of it?

File: robot/framesystem/framesystem.go [hunk 16]
```diff
@@ -131,25 +143,38 @@ func (svc *frameSystemService) Reconfigure(ctx context.Context, deps resource.De
...
+
+	parts, err := svc.ConfigToParts(fsCfg)
+	if err != nil {
+		svc.parts = nil
+		svc.lastReconfigureError = err
 		return err
 	}
 
-	sortedParts, err := referenceframe.TopologicallySortParts(fsCfg.Parts)
+	sortedParts, err := referenceframe.TopologicallySortParts(parts)
```
Comment (biotinker) [inline robot/framesystem/framesystem.go:169]: This will return a `NewParentFrameMissingError` in the case on line 127 above, where one component is not yet ready but perhaps one of its children is.

This should be eventually consistent, but I believe that as written it is now impossible to get even a partial framesystem if any part with children is offline.

Comment (raybjork) [reply robot/framesystem/framesystem.go:169]: Yeah, partial framesystems are never something that we've supported to my knowledge, and this does not attempt to fix this.

**Review by biotinker — CHANGES_REQUESTED** [review]
I have concerns about some of the added `World` checks as it pertains to fetching remote robot framesystems; please set up a local/remote robot, with framesystem members on each, and ensure the local robot includes the properly formed remote framesystem.

If that test passes, the code LGTM

File: referenceframe/frame_json.go (hunk 14 above)
Comment (edaniels) [reply referenceframe/frame_json.go:74]: I think that's covered by https://github.com/viamrobotics/rdk/pull/2415/files#diff-0b975da6011656cd3f2caf0435023573bdb5ac7c6a425e7917dd5cc9d6be63e4R794

File: robot/impl/data/remote_fake.json [hunk 17]
```diff
@@ -36,7 +36,7 @@
         {
             "name": "data_manager1",
             "type": "data_manager",
-            "model": "fake"
+            "model": "builtin"
```
Comment (edaniels) [inline robot/impl/data/remote_fake.json:39]: why did you need this change?

File: referenceframe/frame_system.go [hunk 18]
```diff
@@ -612,28 +594,42 @@ func createFramesFromPart(part *FrameSystemPart) (Frame, Frame, error) {
...
+	// ensure that at least one frame connects to world if the frame system is not empty
+	if len(parts) != 0 {
+		hasWorld := false
+		for _, part := range parts {
+			if part.Origin.Parent() == World {
+				hasWorld = true
+				break
+			}
+		}
+		if !hasWorld {
```
Comment (edaniels) [inline referenceframe/frame_system.go:614]: do you have a test for this that checks for `ErrNoWorldConnection`?

File: referenceframe/frame_system.go [hunk 19]
```diff
@@ -612,28 +594,42 @@ func createFramesFromPart(part *FrameSystemPart) (Frame, Frame, error) {
...
+		names[i] = p.Origin.Name()
 	}
 	return names
 }
 
 // TopologicallySortParts takes a potentially un-ordered slice of frame system parts and
 // sorts them, beginning at the world node.
 func TopologicallySortParts(parts []*FrameSystemPart) ([]*FrameSystemPart, error) {
+	// ensure that at least one frame connects to world if the frame system is not empty
+	if len(parts) != 0 {
```
Comment (edaniels) [inline referenceframe/frame_system.go:606]: if it's empty, should you just return early?

File: robot/framesystem/framesystem.go [hunk 20]
```diff
@@ -279,15 +318,61 @@ func (svc *frameSystemService) TransformPointCloud(ctx context.Context, srcpc po
...
+			geomString,
+		})
+	}
+	return t.Render()
+}
+
+// extractModelFrameJSON finds the robot part with a given name, checks to see if it implements ModelFrame, and returns the
+// JSON []byte if it does, or nil if it doesn't.
+func (svc *frameSystemService) extractModelFrameJSON(name string) (referenceframe.Model, error) {
+	part, ok := svc.components[name]
```
Comment (edaniels) [inline robot/framesystem/framesystem.go:370]: if we're doing some renames, it might be nice to call this `robots` or `robotParts`?

File: referenceframe/frame_json.go [hunk 21]
```diff
@@ -87,6 +94,18 @@ func (cfg *LinkConfig) ParseConfig() (*LinkInFrame, error) {
 	return NewLinkInFrame(cfg.Parent, pose, cfg.ID, geom), nil
 }
 
+// CopyWithNewName makes a deep copy of the config with an ID corresponding to the given name.
+func (cfg *LinkConfig) CopyWithNewName(name string) *LinkConfig {
```
Comment (edaniels) [inline referenceframe/frame_json.go:98]: this needs a unit test

File: referenceframe/frame_json.go [hunk 22]
```diff
@@ -70,6 +71,12 @@ func NewLinkConfig(frame staticFrame) (*LinkConfig, error) {
 
 // ParseConfig converts a LinkConfig into a staticFrame.
 func (cfg *LinkConfig) ParseConfig() (*LinkInFrame, error) {
+	if cfg.ID == World {
+		return nil, errors.Errorf("cannot give frame system part the name %s", World)
```
Comment (edaniels) [inline referenceframe/frame_json.go:75]: this needs a unit test as well

**Review by edaniels — CHANGES_REQUESTED** [review]
basically lgtm but have a few comments

**Comment by benjirewis** [comment]
Reopening + rebasing to see what was failing here. I can no longer repro the error I saw when I filed RSDK-4012.

**Comment by benjirewis** [comment]
😮‍💨 closing again. @raybjork and I discovered that caching with the current system of sensitivity to remote robot reconfiguration may only be a marginal improvement (we still have to rebuild the framesystem fairly often). Something like RSDK-4329 may eventually help.
